src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/ECDiffieHellmanCng.cs

```diff
@@ -30,6 +31,16 @@ public ECDiffieHellmanCng(CngKey key)
             Key = CngAlgorithmCore.Duplicate(key);
         }
 
+        [SupportedOSPlatform("windows")]
+        internal ECDiffieHellmanCng(CngKey key, bool transferOwnership)
+        {
+            Debug.Assert(key is not null);
+            Debug.Assert(key.AlgorithmGroup == CngAlgorithmGroup.ECDiffieHellman);
```

Comment:
@bartonjs does this need to stay as a `if` / `throw` like the public constructors? In the Windows PAL, `DecodeECPublicKey` doesn't know the algorithm group and I think was leaving it to the public constructor to check the algorithm group.

https://github.com/dotnet/runtime/blob/f86e5a0187652e4a9cc59327607093c4807fa5b2/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Pal.Windows.PublicKey.cs#L100-L101

I wonder if this would then let you get pass a `CngKey` with the wrong algorithm group.

For `ECDsa`, we don't care - we take an ECDsa or ECDiffieHellman key because CNG does that.

For ECDiffieHellman, we `CRYPT_OID_INFO_PUBKEY_ENCRYPT_KEY_FLAG`... so I don't think you'd end up passing an ECDsa key in to ECDiffieHellman.

So maybe everything is fine, but I thought it's an interesting scenario to highlight.

Reply:
`set_Key` has the if+throw, so the runtime behavior is still protected.  

https://github.com/dotnet/runtime/blob/0b847f1394fb9565e77273aa3349562c80dc4c9a/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/ECDiffieHellmanCng.Key.cs#L37-L43

Given that we only get here from a certificate that was identified as ECDH we should be good (granted, you can do whatever you want with cert properties, so you could have an ECDH cert with an associated private key that happens to be RSA... but it's nonsense...), so the assert is fine.

The public ctor is probably just doing a bit of looking ahead and avoiding duplicating the key if it's just going to turn around and throw.